File: primo_anno/programmazione_dei_laboratori_con_laboratorio/esami/mio_esame1/scursatone_cristian.c
#include <stdio.h>
#include <stdlib.h>
#include <string.h>
char *str_cat(char *, int, char *);
char *mod(char *);
int main(int n, char *args[]) {
  char *a = malloc(9 * sizeof(char));
  char *b = malloc(9 * sizeof(char));
  for (int i = 0; i < 8; i++) {
    a[i] = 'o';
    b[i] = 'o';
  }
  b[8] = '\0';
  b[8] = '\0';
  printf("%s\n%s\n", a, b);
  int j = 8;
  a = str_cat(a, j, b);
  printf("%s", a);
}

char *str_cat(char *a, int n, char *b) {
  int len_a = strlen(a);
  int len_b = strlen(b);
  if (n < len_a + len_b + 1) {
    n = len_b + len_a + 1;
    a = realloc(a, n * sizeof(char));
  }
  for (int i = 0; i < len_b; i++) {
    a[len_a + i] = b[i];
  }
  a[n - 1] = '\0';
  free(b);

  return a;
}
/*
 * len  a =n
 * len b = m
 * costo temporale = O(n+m) nel caso peggiore,se bisogna fare il realloc, O(m)
 * nel caso migliore,
 *
 * costo spaziale = O(1)*/

File: primo_anno/programmazione_dei_laboratori_con_laboratorio/esami/mio_esame1/scursatone_cristian.py
import math


def mode(a):
    d = {}
    l = []
    for i in a:
        d[i] = d.get(i, 0) + 1
    mass = max(d.values())
    for i in d.keys():
        if d[i] == mass:
            l.append(i)
    return l


a = [3, 1, 1, 1, 1, 2, 2, 2, 2, 3, 3]
l = mode(a)
print(l)
# n = len(a)
# costo temporale = O(n+k), con k = d.keys(), ma nel caso peggiore è uguale a n, quindi O(n+O(n))= O(n)
# costo spaziale = O(k) con k = d.keys(), nel caso peggiore è n, quindi O(n) nel caso peggiore
